modules/core/04-channel/keeper/upgrade.go

```diff
@@ -131,125 +96,98 @@ func (k Keeper) ChanUpgradeTry(ctx sdk.Context, portID string, channelID string,
 	if err := k.connectionKeeper.VerifyChannelState(ctx, connectionEnd, proofHeight, proofChannel, channel.Counterparty.PortId,
-		channel.Counterparty.ChannelId, counterpartyUpgradeChannel); err != nil {
+		channel.Counterparty.ChannelId, expectedCounterpartyChannel); err != nil {
 		return 0, err
 	}
 
-	upgradeTimeout := types.UpgradeTimeout{TimeoutHeight: timeoutHeight, TimeoutTimestamp: timeoutTimestamp}
-	if err := k.connectionKeeper.VerifyChannelUpgradeTimeout(ctx, connectionEnd, proofHeight, proofUpgradeTimeout, channel.Counterparty.PortId,
-		channel.Counterparty.ChannelId, upgradeTimeout); err != nil {
+	// verify the proposed upgrade information provided by the counterparty
+	if err := k.connectionKeeper.VerifyChannelUpgrade(ctx, connectionEnd, proofHeight, proofUpgrade, channel.Counterparty.PortId,
+		channel.Counterparty.ChannelId, counterpartyProposedUpgrade); err != nil {
 		return 0, err
 	}
 
-	if err := k.connectionKeeper.VerifyChannelUpgradeSequence(ctx, connectionEnd, proofHeight, proofUpgradeSequence, channel.Counterparty.PortId,
-		channel.Counterparty.ChannelId, counterpartyUpgradeSequence); err != nil {
-		return 0, err
-	}
+	/*
+		replace the following with:
+
+		selfHeight := clienttypes.GetSelfHeight(ctx)
+		selfTime := uint64(ctx.BlockTime.UnixNano())
+		if info, hasPassed := counterpartyProposedUpgrade.Timeout.Status(selfHeight, selfTime); hasPassed {
+			return 0, types.NewErrorReceipt(upgradeSequence, info)
+		}
+	*/
 
 	// check if upgrade timed out by comparing it with the latest height of the chain
 	selfHeight := clienttypes.GetSelfHeight(ctx)
+	timeoutHeight := counterpartyProposedUpgrade.Timeout.TimeoutHeight
 	if !timeoutHeight.IsZero() && selfHeight.GTE(timeoutHeight) {
-		if err := k.RestoreChannelAndWriteErrorReceipt(ctx, portID, channelID, upgradeSequence, types.ErrUpgradeTimeout); err != nil {
+		if err := k.WriteErrorReceipt(ctx, portID, channelID, upgradeSequence, types.ErrUpgradeTimeout); err != nil {
 			return 0, errorsmod.Wrap(types.ErrUpgradeAborted, err.Error())
 		}
 		return 0, errorsmod.Wrapf(types.ErrUpgradeAborted, "block height >= upgrade timeout height (%s >= %s)", selfHeight, timeoutHeight)
 	}
 
 	// check if upgrade timed out by comparing it with the latest timestamp of the chain
+	timeoutTimestamp := counterpartyProposedUpgrade.Timeout.TimeoutTimestamp
 	if timeoutTimestamp != 0 && uint64(ctx.BlockTime().UnixNano()) >= timeoutTimestamp {
 		upgradeSequence = uint64(0)
-		if err := k.RestoreChannelAndWriteErrorReceipt(ctx, portID, channelID, upgradeSequence, types.ErrUpgradeTimeout); err != nil {
+		if err := k.WriteErrorReceipt(ctx, portID, channelID, upgradeSequence, types.ErrUpgradeTimeout); err != nil {
 			return 0, errorsmod.Wrap(types.ErrUpgradeAborted, err.Error())
 		}
 		return 0, errorsmod.Wrapf(types.ErrUpgradeAborted, "block timestamp >= upgrade timeout timestamp (%s >= %s)", ctx.BlockTime(), time.Unix(0, int64(timeoutTimestamp)))
 	}
 
-	switch channel.State {
```

Comment:
I found this switch to being doing several unrelated pieces of logic thus making it hard to understand. This was my understanding of what it was trying to accomplish (and the change I made)

```go
if channel == OPEN {
    // increment upgrade sequence appropriately
}

if upgradeSequence != counterpartyUpgradeSequence {
    // abort
    // write error receipt
    
    if channel.State == INITUPGRADE {
        // in crossing hello case we can fast forward sequences
        fastForward(ctx, channel, counterpartyUpgradeSequence)
    }
    
    return ErrorReceipt
}
```
   

```